config add: use the provided value to validate the type

chuckatkins opened the pull request:
Fixes #17543
Fixes #23259

When adding a config parameter via `spack config add 'foo:bar:baz'` an attempt is made to determine the appropriate type for the config entry by cycling through the default values for various types and attempting to validate against the schema.  This fails for enums since the default value is most likely not one of the valid enum values.  This change uses the user-provided value itself to construct the test value to validate with rather than the default value for a given type.

Before this change the following would fail and now succeeds:

```bash
$ spack config add 'config:concretizer:clingo'
==> Error: Cannot determine valid type for path 'config:concretizer'.
```

Commit:
config: When adding a config entry, use the provided value to validate the type

When adding a config parameter via "spack config add foo:bar:baz" an attempt is made to determine
the appropriate type for the config entry by cycling through the default values for various types
and attempting to validate against the schema. This fails for enums since the default value is most
likely not one of the valid enum values. This change uses the user-provided value itself to
construct the test value to validate with rather than the default value for a given type.

Before this change the following would fail and now succeeds:

$ spack config add config:concretizer:clingo
==> Error: Cannot determine valid type for path 'config:concretizer'.

chuckatkins commented:
Making the style check happy for `config.py` seems to put the imports in a weird order:
```diff
diff --git a/lib/spack/spack/config.py b/lib/spack/spack/config.py
index 18d8ff59e7..fb72056e34 100644
--- a/lib/spack/spack/config.py
+++ b/lib/spack/spack/config.py
@@ -36,35 +36,32 @@
 import re
 import sys
 from contextlib import contextmanager
-from six import iteritems
-from ordereddict_backport import OrderedDict
 from typing import List  # novm
 
 import ruamel.yaml as yaml
+from ordereddict_backport import OrderedDict
 from ruamel.yaml.error import MarkedYAMLError
+from six import iteritems
 
 import llnl.util.lang
 import llnl.util.tty as tty
-from llnl.util.filesystem import mkdirp
-
-import spack.paths
 import spack.architecture
 import spack.compilers
+import spack.paths
 import spack.schema
 import spack.schema.compilers
+import spack.schema.config
+import spack.schema.env
 import spack.schema.mirrors
-import spack.schema.repos
-import spack.schema.packages
 import spack.schema.modules
-import spack.schema.config
+import spack.schema.packages
+import spack.schema.repos
 import spack.schema.upstreams
-import spack.schema.env
+import spack.util.spack_yaml as syaml
+from llnl.util.filesystem import mkdirp
 from spack.error import SpackError
 from spack.util.cpus import cpus_available
 
-# Hacked yaml for configuration files preserves line numbers.
-import spack.util.spack_yaml as syaml
-
 #: Dict from section names -> schema for that section
 section_schemas = {
     'compilers': spack.schema.compilers.schema,
```

Is there a better way to make that check happy?

alalazo commented:
@chuckatkins Can you please rebase this one? All imports in spack have been sorted now.

tgamblin committed:
config: add test for type validation; fix mutated test config

- [x] add a test for using enumerated types in the tests for `spack config add`

- [x] make `config add` tests use the `mutable_config` fixture so they do not
      affect other tests

tgamblin commented:
@chuckatkins @becker33: I added a test for this and fixed a bug in the tests where we were not using the `mutable_config` fixture.

tgamblin commented:
for there record, @chuckatkins is going to look at this.  The issues seems to be what to do for lists in paths.

tgamblin commented:
@chuckatkins did you get a chance to check this one out?  I can also look into it again.

haampie commented:
This PR also solves my problem of not being able to change numeric values... would be great to get this in.

```
$ spack config add config:build_jobs:123
==> Error: Cannot determine valid type for path 'config:build_jobs'.
```

tgamblin commented:
Agree.

@chuckatkins was going to look at it but looks like one of us would need to.  Current issue is that this fails on list values. If someone wants to fix that it would be great.  Otherwise I will eventually get to it.

haampie commented:
Moving this to a next spack release then...
